File: sdk/elixir/lib/telecom_sdk/http_client.ex
defmodule TelecomSDK.HTTPClient do
  @moduledoc """
  HTTP client for making API requests
  """

  defstruct [:base_url, :auth_provider, :timeout, :max_retries, :retry_delay, :enable_logging]

  def new(base_url, auth_provider, opts \\ []) do
    %__MODULE__{
      base_url: base_url,
      auth_provider: auth_provider,
      timeout: Keyword.get(opts, :timeout, 30_000),
      max_retries: Keyword.get(opts, :max_retries, 3),
      retry_delay: Keyword.get(opts, :retry_delay, 1_000),
      enable_logging: Keyword.get(opts, :enable_logging, false)
    }
  end

  def get(client, path, params \\ %{}) do
    request(client, "GET", path, nil, params)
  end

  def post(client, path, body \\ nil) do
    request(client, "POST", path, body, nil)
  end

  def put(client, path, body \\ nil) do
    request(client, "PUT", path, body, nil)
  end

  def delete(client, path) do
    request(client, "DELETE", path, nil, nil)
  end

  defp request(client, method, path, body, params) do
    url = client.base_url <> path
    headers = TelecomSDK.AuthProvider.get_headers(client.auth_provider)

    opts = [
      timeout: client.timeout,
      recv_timeout: client.timeout
    ]

    opts = if params != %{}, do: Keyword.put(opts, :params, params), else: opts
    opts = if body, do: Keyword.put(opts, :body, Jason.encode!(body)), else: opts

    last_error = nil

    Enum.reduce_while(0..client.max_retries, nil, fn attempt, _acc ->
      case do_request(method, url, headers, opts) do
        {:ok, response} ->
          case handle_response_errors(response) do
            :ok -> {:halt, {:ok, Jason.decode!(response.body)}}
            {:error, error} -> {:halt, {:error, error}}
          end
        {:error, error} ->
          if attempt < client.max_retries do
            Process.sleep(client.retry_delay * round(:math.pow(2, attempt)))
            {:cont, error}
          else
            {:halt, {:error, error}}
          end
      end
    end)
  end

  defp do_request(method, url, headers, opts) do
    case HTTPoison.request(method, url, Keyword.get(opts, :body), headers, opts) do
      {:ok, response} -> {:ok, response}
      {:error, error} -> {:error, error}
    end
  end

  defp handle_response_errors(%{status_code: status}) when status >= 200 and status < 300, do: :ok
  defp handle_response_errors(%{status_code: 401}), do: {:error, :authentication_failed}
  defp handle_response_errors(%{status_code: 429}), do: {:error, :rate_limit_exceeded}
  defp handle_response_errors(%{status_code: status}) when status >= 400 and status < 500, do: {:error, {:client_error, status}}
  defp handle_response_errors(%{status_code: status}) when status >= 500, do: {:error, {:server_error, status}}

  def close(_client) do
    # HTTPoison doesn't need explicit cleanup
    :ok
  end
end

File: sdk/ruby/lib/telecom_sdk/http_client.rb
require "httparty"

module TelecomSDK
  # HTTP client for making API requests
  class HTTPClient
    include HTTParty

    def initialize(base_url, auth_provider, timeout: 30, max_retries: 3, retry_delay: 1, enable_logging: false)
      @base_url = base_url
      @auth_provider = auth_provider
      @timeout = timeout
      @max_retries = max_retries
      @retry_delay = retry_delay
      @enable_logging = enable_logging
    end

    def get(path, params = {})
      request("GET", path, nil, params)
    end

    def post(path, body = nil)
      request("POST", path, body, nil)
    end

    def put(path, body = nil)
      request("PUT", path, body, nil)
    end

    def delete(path)
      request("DELETE", path, nil, nil)
    end

    private

    def request(method, path, body, params)
      url = "#{@base_url}#{path}"
      headers = @auth_provider.get_headers

      options = {
        timeout: @timeout,
        headers: headers
      }

      options[:query] = params if params
      options[:body] = body.to_json if body

      last_error = nil
      (0..@max_retries).each do |attempt|
        begin
          response = HTTParty.send(method.downcase, url, options)
          handle_response_errors(response)
          return response.parsed_response
        rescue StandardError => e
          last_error = e
          if attempt < @max_retries
            sleep(@retry_delay * (2 ** attempt))
          end
        end
      end

      raise last_error || NetworkError, "Request failed after #{@max_retries} retries"
    end

    def handle_response_errors(response)
      case response.code
      when 401
        raise AuthenticationError, "Authentication failed"
      when 429
        raise RateLimitError, "Rate limit exceeded"
      when 400..499
        raise APIError, "API error: #{response.parsed_response}"
      when 500..599
        raise ServerError, "Server error: #{response.code}"
      end
    end

    def close
      # HTTParty doesn't need explicit cleanup
    end
  end
end
